### ISSUE TITLE
Scheduled activities sometimes fail to start causing orchestration to hang

### ISSUE THREAD

**Issue by starkna** [Azure/azure-functions-durable-extension#822]
As described in #780 - Starting new issue report specific to this problem. 

We have occasional orchestrations running where a number of activities are scheduled, but some of those activities are never started/completed, causing the orchestrator to hang and never complete.

- Durable Functions extension version:  1.8.2
- Function App version (1.0 or 2.0):  2.0
- Programming language used:  C#

- Timeframe issue observed:
- Function App name:  DdsMfgDataGeneration-Research
- Function name(s):  MfgDataGenerationFunction
- Region:  CentralUS
- Orchestration instance ID(s):  c94f33f614544bc78223eae78b0d5ed8

**Comment by cgillum** [comment]
Sorry for the delay on this. From what I can observe, it looks like this orchestration instance actually completed. It started at **2019-06-25 21:04:35** and completed at **2019-06-26 16:21:20** (both times UTC).

I did notice that it failed with an exception:

```
Microsoft.Azure.WebJobs.FunctionFailedException: The activity function 'CompleteWork' failed: "CompleteWorkAsync failed; model is no longer checked out to current user.". See the function execution logs for additional details. ---> Mmm.Ocsd.Dds.Common.Exceptions.ModelNotCheckedOutException: CompleteWorkAsync failed; model is no longer checked out to current user.
   --- End of inner exception stack trace ---
   at Microsoft.Azure.WebJobs.DurableOrchestrationContext.CallDurableTaskFunctionAsync[TResult](String functionName, FunctionType functionType, String instanceId, RetryOptions retryOptions, Object input) in E:\functions\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\DurableOrchestrationContext.cs:line 498
   at Mmm.Ocsd.Dds.Functions.MfgDataGeneration.MfgDataGenerationFunction.RunOrchestrator(DurableOrchestrationContext context, ILogger log) in D:\a\14\s\src\Mmm.Ocsd.Dds.Functions.MfgDataGeneration\MfgDataGenerationFunction.cs:line 140
   at Microsoft.Azure.WebJobs.Host.Executors.VoidTaskMethodInvoker`2.InvokeAsync(TReflected instance, Object[] arguments) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\VoidTaskMethodInvoker.cs:line 20
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync(Object instance, Object[] arguments) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs:line 52
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.TaskOrchestrationShim.Execute(OrchestrationContext innerContext, String serializedInput) in E:\functions\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\Listener\TaskOrchestrationShim.cs:line 91
```

I also noticed that the durable timer that was scheduled for this orchestration actually fired at a significantly later time (2019-06-26 14:16:35) than when it was scheduled for (2019-06-25T23:04:37). I did not look into this closely, however. By chance was your application stopped for some period of time?

**Comment by starkna** [comment]
We've re-organized our storage accounts since this time, so I can't find the logs anymore for this particular issue to refer back to for more details.  From my recollection, though, there were somewhere on the order of 40 fan-out activities that were scheduled by the orchestrator, but only some portion of them (between 30 and 40) that actually started.  After about a day of waiting for something to happen with this orchestrator, I tried restarting the function app.  After the restart, I believe the timer finally fired (the timer was scheduled to fire after 2 hours), and the orchestration completed because of this.  

I updated to the 1.8.3 version of the Durable Function extension, and I have not been seeing this issue in the last few weeks, so perhaps it has been resolved by some fix in that release.  We have been running all of this in our testing environments for the past couple of months, and we will be releasing it all into production next month, so once it is under heavier load, we'll be able to determine if this is still an issue for us or not.  If you want to go ahead and close the issue for now, I can open a new issue if that does happen.  Thanks.

**Comment by starkna** [comment]
@cgillum We have deployed to production, and we have run into several issues on one of our Azure Functions (while others have all been working that are doing similar fanout/aggregation tasks).  I have a case in a staging environment that shows what we are struggling with:

Azure Function App: DdsMfgDataGeneration-SL1-Staging
Function Name: MfgDataGenerationFunction
Execution Id: 588c347a94b74d81b3c7b9a6fb92d15a

Running this function, we scheduled 7 activity tasks on a fanout.  If you look at the 7 TaskCompleted events that were processed, 2 of them were received for TaskScheduledId 3 and none were received for TaskScheduledId 5.  It seems that somehow if multiple results are received simulataneously under some conditions, they are not being processed correctly in regards to their TaskId.  This causes the orchestration to wait for the remainder of the tasks (TaskId 5).  I added a timer after 15 minutes to try to re-submit any jobs that have not completed, which does re-send the task that did not complete for re-execution.  However, my concern is with the 2 separate tasks that came back, both being processed as Task Id 3 results - if one of them is not actually from task 3, then I could be writing invalid results back to the wrong object in the aggregation.
